File: 2016/day2.js
let input = require('./input/day2')
var input2 = `ULL
RRDDD
LURDL
UUUUD`

var keypad = [
 [1, 2, 3],
 [4, 5, 6],
 [7, 8, 9]
]
var complexKeypad = [
  [null, null, 1],
  [null, 2,    3,   4],
  [5,    6,    7,   8,  9],
  [null, 'A',  'B', 'C'],
  [null, null, 'D']
]
var position = [2, 0]
var lines = input.split("\n")

var code = lines.map(lineToPosition).map(positionToNumber)
console.log(code.join(''))

function lineToPosition (line) {
  console.log('newline')
  let letters = line.split('')
  return letters.reduce(lettersToNumber, position)
}
function lettersToNumber (pos, move) {
  switch (move) {
    case 'U':
      console.log('U')
      position = upFrom(pos)
      break;
    case 'D':
      console.log('D')
      position = downFrom(pos)
      break;
    case 'R':
      console.log('R')
      position = rightFrom(pos)
      break;
    case 'L':
      console.log('L')
      position = leftFrom(pos)
      break;
  }
  console.log('move ', move, 'to pos ', position, positionToNumber(position))
  return position
}
function upFrom (position) {
  return move(position, -1, 0)
}
function move(position, vertical, horizontal) {
  if (position[0] + vertical < 0 || position[1] + horizontal < 0
     || position[0] + vertical > 4)
    return position
  var newNumber = complexKeypad[position[0] + vertical][position[1] + horizontal]
  if (newNumber){
    return [position[0] + vertical, position[1] + horizontal]}
  return position
}
function downFrom (position) {
  return move(position, 1, 0)
}
function rightFrom (position) {
  return move(position, 0, 1)
}
function leftFrom (position) {
  return move(position, 0, -1)
}
function positionToNumber (position) {
  return complexKeypad[position[0]][position[1]]
}

File: 2016/day2.rb
input = File.read('./input/day2.txt')

keypad = "123
456
789"

test = "ULL
RRDDD
LURDL
UUUUD"

def move(row, col, keypad, char)
  if char == 'U'
    attempt_row, attempt_col = row - 1, col
  elsif char == 'L'
    attempt_row, attempt_col = row, col - 1
  elsif char == 'R'
    attempt_row, attempt_col = row, col + 1
  elsif char == 'D'
    attempt_row, attempt_col = row + 1, col
  end
  if attempt_row >= 0 && 
      attempt_col >= 0 && 
      keypad[attempt_row] && ![nil, " "].include?(keypad[attempt_row][attempt_col])
    return [attempt_row, attempt_col]
  else
    return [row, col]
  end
end

def part1(instructions, keypad)
  instructions = instructions.split("\n").map{|line| line.split("")}
  keypad = keypad.split("\n").map{|line| line.split("")}
  row = keypad.index(keypad.find{|row| row.include?("5")})
  col = keypad[row].index("5")
  code = ""
  instructions.each do |chars|
    chars.each do |char|
      row, col = move(row, col, keypad, char)
    end
    code += keypad[row][col]
  end
  code
end

puts part1(test, keypad)
puts part1(input, keypad)

badKeypad = "  1  
 234 
56789
 ABC 
  D"
puts part1(test, badKeypad)
puts part1(input, badKeypad)
